src/hardware/vga_draw.cpp

```diff
@@ -351,6 +352,47 @@ static uint8_t * VGA_Draw_Linear_Line(Bitu vidstart, Bitu /*line*/) {
 	return ret;
 }
 
+static uint8_t* draw_unwrapped_line_from_dac_palette(Bitu vidstart, Bitu)
+{
+	// Quick references
+	static constexpr auto palette_map        = vga.dac.palette_map;
+	static constexpr uint8_t bytes_per_pixel = sizeof(palette_map[0]);
+	static const auto linear_mask            = vga.draw.linear_mask;
+	static const auto linear_addr            = vga.draw.linear_base;
+
+	// Video mode-specific line variables
+	const auto pixels_in_line = static_cast<uint16_t>(vga.draw.line_length /
+	                                                  bytes_per_pixel);
+	const auto video_end = vidstart + pixels_in_line;
+
+	// The line address is where the RGB888 palettized pixel is written.
+	// It's incremented forward per pixel.
+	auto line_addr = reinterpret_cast<uint32_t*>(TempLine);
+
+	auto linear_pos = vidstart;
+
+	// This function typically runs on 640+-wide lines and is a rendering
+	// bottleneck. Draw in batches of four to let the host pipeline deeper.
+	while (linear_pos < video_end) {
+		const auto masked_pos_0 = linear_pos++ & linear_mask;
+		const auto masked_pos_1 = linear_pos++ & linear_mask;
+		const auto masked_pos_2 = linear_pos++ & linear_mask;
+		const auto masked_pos_3 = linear_pos++ & linear_mask;
+
+		const auto palette_index_0 = *(linear_addr + masked_pos_0);
+		const auto palette_index_1 = *(linear_addr + masked_pos_1);
+		const auto palette_index_2 = *(linear_addr + masked_pos_2);
+		const auto palette_index_3 = *(linear_addr + masked_pos_3);
+
+		*line_addr++ = *(palette_map + palette_index_0);
+		*line_addr++ = *(palette_map + palette_index_1);
+		*line_addr++ = *(palette_map + palette_index_2);
+		*line_addr++ = *(palette_map + palette_index_3);
+	}
```

Comment:
No need to manually unroll; GCC is more than clever enough to do that if needed. Just do something like this:

```cpp
auto num_repeats = 4;
while (num_repeats--) {
   ...
}
```

Reply:
Plus perhaps also assert that line length MOD 4 equals zero (mostly for documentation purposes).

Reply:
The inner loop still unfortunately isn't being vectorized :sob: 

### GCC

![2023-05-17_19-38_1](https://github.com/dosbox-staging/dosbox-staging/assets/1557255/ab8c4946-154a-464e-ad8e-8c5df6536eff)

### Clang

![2023-05-17_19-38](https://github.com/dosbox-staging/dosbox-staging/assets/1557255/906c16a9-e4ac-4025-9e7c-994bee5063de)

---

I included the while-loop in the last commit, however comparing it to the explicit version, I get about 2 to 3 fps more in QTG 640x480.

### While-loop

![2023-05-17_19-42_1](https://github.com/dosbox-staging/dosbox-staging/assets/1557255/f6e0209d-4d8d-4082-943b-882a8e816305)
![2023-05-17_19-40_1](https://github.com/dosbox-staging/dosbox-staging/assets/1557255/596335f8-7ed9-43b7-afe8-2f4bc7be7467)

### Without

![2023-05-17_19-45](https://github.com/dosbox-staging/dosbox-staging/assets/1557255/94c14dfd-29a7-479b-9d10-feb9845a0736)
![2023-05-17_19-44_1](https://github.com/dosbox-staging/dosbox-staging/assets/1557255/a86abf15-f9e6-47f1-8021-0fd8b1371e61)

Reply:
That's... disappointing. Stupid GCC 😢 Maybe leave the explicit unrolling then 🤷🏻

Reply:
Yup - will do a comparison on the pi where it's the quietest environment and any improvement has the most benefit.

Reply:
On a second thought... There's nothing to vectorise here, we're not doing any calculations, just 4 random table lookups. Of course, four memory reads from random memory locations cannot be vectorised!

Also, I _highly_ doubt the explicit unrolling vs the loop approach would make such a huge FPS difference (huge compared to how computationally infinitesimally, microscopically small incrementing a counter in a loop is...) Also, I doubt any unrolling would make any difference here at all. Have you measured that doing lookups 4 times here improves anything, or was it a guess?

Reply:
The `while` loop won all three rounds on the Pi:

| round | neither | unrolled | while |
|:---:|:---:|:---:|:---:|
1 | 6.4 | 6.8 | 7.0 
2 | 6.5 | 6.9 | 7.2 
3 | 6.6 | 6.8 | 7.1
 

Ran in round-robin all back-to-back. Planning to leave the PR as-is.

> Of course, four memory reads from random memory locations cannot be vectorised!

I think a scatter gather approach would work - the initial masked pos values are (usually) adjacent, as are the writes at the end (so those could maybe be SIMD'd) - with a scatter/gather in the middle.

But yeah.. pretty bandwidth-sensitive area .. these frame-rates are s slow, you can watch the pi struggle to grind through these lookups :sweat_smile:

Reply:
Nice measurements @kcgen , the `while` loop variant it is then 🚀 Thanks for testing this!